modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,470 @@
+  [_ ^ZonedDateTime t]
+  ;; for native query parameter substitution, in order to not conflict with the `PrestoConnection` session time zone
+  ;; (which was set via report time zone), it is necessary to use the `from_iso8601_timestamp` function on the string
+  ;; representation of the `ZonedDateTime` instance, but converted to the report time zone
+  (let [report-zone       (qp.timezone/report-timezone-id-if-supported :presto-jdbc)
+        ^ZonedDateTime ts (if (str/blank? report-zone) t (t/with-zone-same-instant t (t/zone-id report-zone)))]
+    ;; the `from_iso8601_timestamp` only accepts timestamps with an offset (not a zone ID), so only format with offset
+    (date-time->substitution (.format ts DateTimeFormatter/ISO_OFFSET_DATE_TIME))))
+
+(defmethod sql.params.substitution/->prepared-substitution [:presto-jdbc LocalDateTime]
+  [_ ^LocalDateTime t]
+  ;; similar to above implementation, but for `LocalDateTime`
+  ;; when Presto parses this, it will account for session (report) time zone
+  (date-time->substitution (.format t DateTimeFormatter/ISO_LOCAL_DATE_TIME)))
+
+(defn- set-time-param
+  "Converts the given instance of `java.time.temporal`, assumed to be a time (either `LocalTime` or `OffsetTime`)
+  into a `java.sql.Time`, including milliseconds, and sets the result as a parameter of the `PreparedStatement` `ps`
+  at index `i`."
+  [^PreparedStatement ps ^Integer i ^Temporal t]
+  ;; for some reason, `java-time` can't handle passing millis to java.sql.Time, so this is the most straightforward way
+  ;; I could find to do it
+  ;; reported as https://github.com/dm3/clojure.java-time/issues/74
+  (let [millis-of-day (.get t ChronoField/MILLI_OF_DAY)]
+    (.setTime ps i (Time. millis-of-day))))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetTime]
+  [_ ^PreparedStatement ps ^Integer i t]
+  ;; necessary because `PrestoPreparedStatement` does not implement the `setTime` overload having the final `Calendar`
+  ;; param
+  (let [adjusted-tz (t/with-offset-same-instant t (t/zone-offset 0))]
+    (set-time-param ps i adjusted-tz)))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc LocalTime]
+  [_ ^PreparedStatement ps ^Integer i t]
+  ;; same rationale as above
+  (set-time-param ps i t))
+
+(defn- ^LocalTime sql-time->local-time
+  "Converts the given instance of `java.sql.Time` into a `java.time.LocalTime`, including milliseconds. Needed for
+  similar reasons as `set-time-param` above."
+  [^Time sql-time]
+  (LocalTime/ofInstant (Instant/ofEpochMilli (.getTime sql-time)) (t/zone-id "UTC")))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:presto-jdbc Types/TIME]
+  [_ ^ResultSet rs ^ResultSetMetaData rs-meta ^Integer i]
+  #(let [local-time (-> (.getTime rs i)
+                        sql-time->local-time)
+         type-name  (.getColumnTypeName rs-meta i)
+         base-type  (presto-common/presto-type->base-type type-name)]
```

Comment:
You should do the `type-name` and `base-type` and `isa?` stuff _once_, and then return an appropriate thunk function based on that. Otherwise we're potentially do that calculation millions of times... with the multimethod dispatch and the fact that `presto-type->base-type` does a linear comparison against a bunch of regexes), that overhead is definitely going add up. E.g. if you have a million-row result set with 3 TIME columns = 3 million calculations (instead of just doing 3, one for each column)

(That's the point of `read-column-thunk` -- return a minimal thunk function that can be used to fetch the value for a given column in the current row of the result set).